File: src/rotary/rotary.cxx
/*
 * Authors: Brendan Le Foll <brendan.le.foll@intel.com>
 *          Mihai Tudor Panu <mihai.tudor.panu@intel.com>
 *          Sarah Knepper <sarah.knepper@intel.com>
 * Copyright (c) 2014 - 2016 Intel Corporation.
 *
 * This program and the accompanying materials are made available under the
 * terms of the The MIT License which is available at
 * https://opensource.org/licenses/MIT.
 *
 * SPDX-License-Identifier: MIT
 */

#include <iostream>
#include <string>
#include <stdexcept>

#include "rotary.hpp"
#include "math.h"

using namespace upm;

Rotary::Rotary(unsigned int pin)
{
    if ( !(m_aio = mraa_aio_init(pin)) ) {
        throw std::invalid_argument(std::string(__FUNCTION__) +
                                    ": mraa_aio_init() failed, invalid pin?");
        return;
    }
}

Rotary::~Rotary()
{
    mraa_aio_close(m_aio);
}

float Rotary::abs_value()
{
    return (float) mraa_aio_read(m_aio);
}

float Rotary::abs_deg()
{
    return Rotary::abs_value() * (float) m_max_angle / 1023.0;
}

float Rotary::abs_rad()
{
    return Rotary::abs_deg() * M_PI / 180.0;
}

float Rotary::rel_value()
{
    return Rotary::abs_value() - 512.0;
}

float Rotary::rel_deg()
{
    return Rotary::rel_value() * (float) m_max_angle / 1023.0;
}

float Rotary::rel_rad()
{
    return Rotary::rel_deg() * M_PI / 180.0;
}

 float Rotary::getAngle()
 {
     return Rotary::abs_deg();
 }

File: examples/javascript/rotary.js
/*
 * Author: Mihai Tudor Panu <mihai.tudor.panu@intel.com>
 * Copyright (c) 2014 Intel Corporation.
 *
 * This program and the accompanying materials are made available under the
 * terms of the The MIT License which is available at
 * https://opensource.org/licenses/MIT.
 *
 * SPDX-License-Identifier: MIT
 */

//initialization
var upm = require('jsupm_rotary');

//setup access analog input Analog pin #0 (A0)
var rotary = new upm.Rotary(0);

loop();

function loop() {
    var abs = rotary.abs_value();
    var absdeg = rotary.abs_deg();
    var absrad = rotary.abs_rad();

    var rel = rotary.rel_value();
    var reldeg = rotary.rel_deg();
    var relrad = rotary.rel_rad();

    //write the knob value to the console in different formats
    console.log("Abs: " + abs + " " + Math.round(parseInt(absdeg)) + " " + absrad.toFixed(3));
    console.log("Rel: " + rel + " " + Math.round(parseInt(reldeg)) + " " + relrad.toFixed(3));

    //wait 2s and call function again
    setTimeout(loop, 2000);
}
